Cabal/Distribution/Simple/Configure.hs

```diff
@@ -1208,21 +1208,24 @@ selectDependency pkgid internalIndex installedIndex requiredDepsMap
                     else do_internal
     _          -> do_external Nothing
   where
-    do_internal = Right (InternalDependency dep
-                    (PackageIdentifier dep_pkgname (packageVersion pkgid)))
-    do_external is_internal = case Map.lookup dep_pkgname requiredDepsMap of
-      -- If we know the exact pkg to use, then use it.
-      Just pkginstance -> Right (ExternalDependency dep pkginstance)
-      -- Otherwise we just pick an arbitrary instance of the latest version.
-      Nothing -> case PackageIndex.lookupDependency installedIndex dep' of
-        []   -> Left  $
-                  case is_internal of
-                    Just cname -> DependencyMissingInternal dep_pkgname
-                                    (computeCompatPackageName (packageName pkgid) cname)
-                    Nothing -> DependencyNotExists dep_pkgname
-        pkgs -> Right $ ExternalDependency dep $
-                case last pkgs of
-                  (_ver, pkginstances) -> head pkginstances
+    do_internal = Right $ InternalDependency
+                    $ PackageIdentifier dep_pkgname $ packageVersion pkgid
+    do_external is_internal = do
+      ipi <- case Map.lookup dep_pkgname requiredDepsMap of
+        -- If we know the exact pkg to use, then use it.
+        Just pkginstance -> Right pkginstance
+        -- Otherwise we just pick an arbitrary instance of the latest version.
+        Nothing -> case PackageIndex.lookupDependency installedIndex dep' of
+          []   -> Left $ case is_internal of
+            Just cname -> DependencyMissingInternal dep_pkgname
+                   (computeCompatPackageName (packageName pkgid) cname)
+            Nothing -> DependencyNotExists dep_pkgname
+          pkgs -> Right $ head $ snd $ last pkgs
+      -- Fix metadata that may be stripped by old ghc-pkg
```

Comment:
Hmmm... do you rely on this? I'm worried that not all instances of the IPI floating around will be filled in correctly.